### PR TITLE
fix eglib on AIX/PASE

### PR DESCRIPTION
**Pull request by NattyNarwhal**
* Work around access(2) syscall bug on PASE
	  On PASE, X_OK will always returns true. This is contrary to
	  how AIX behaves (though corresponds to documentation!). This
	  is probably because it's an upcall to the ILE version of
	  access, and that seemingly always returns true.

	* Always export the dummy function in the test suite
	  The AIX linker didn't export this properly, without using the
	  giant sledgehammer of -bexpall. Write an export file and use
	  it. This required some build system changes to expose an
	  AIX-only automake conditional.

	* Use a file that exists on PASE for tests
	  PASE doesn't use Unix-style DNS resolution, so /etc/hosts is
	  missing. Use /etc/magic, since that seems to be present.

	* Don't use /bin on AIX/PASE for tests
	  This is a symlink to /usr/bin, and it seems chdir/getcwd
	  follows links. Use /usr instead, that is unlikely to be a
	  symlink.

	* Make a note what directory we're in if we fail test_cwd

### CODE REVIEW

File: mono/eglib/gfile-unix.c [hunk 1]
```diff
@@ -55,8 +55,26 @@ g_file_test (const gchar *filename, GFileTest test)
...
+		 * such, workaround it.
+		 */
+		if (!have_stat)
+			have_stat = (stat (filename, &st) == 0);
+		if (have_stat && (st.st_mode & S_IXUSR) && (st.st_uid == getuid()))
+			return TRUE;
+		if (have_stat && (st.st_mode & S_IXGRP) && (st.st_gid == getgid()))
+			return TRUE;
+		if (have_stat && (st.st_mode & S_IXOTH))
+			return TRUE;
```
Comment (jaykrell) [inline mono/eglib/gfile-unix.c]: This check is cheaper than the previous two. Do it before them?

Comment (jaykrell) [reply mono/eglib/gfile-unix.c]: For that matter, nest everything inside `if (have_stat)`. I realize it perfs the same, but would be cleaner.

Comment (jaykrell) [reply mono/eglib/gfile-unix.c]:
```
	have_stat = have_stat || (stat (filename, &st) == 0); 

	if (have_stat) {
		if ((st.st_mode & S_IXOTH)
				|| ((st.st_mode & S_IXUSR) && (st.st_uid == getuid()))
				|| ((st.st_mode & S_IXGRP) && (st.st_gid == getgid()))
			return TRUE;
```

Comment (jaykrell) [inline mono/eglib/gfile-unix.c]:
```
	have_stat = have_stat || (stat (filename, &st) == 0); 

	if (have_stat && (
		((st.st_mode & S_IXOTH)
			|| ((st.st_mode & S_IXUSR) && (st.st_uid == getuid()))
			|| ((st.st_mode & S_IXGRP) && (st.st_gid == getgid())))))
		return TRUE;
```

File: mono/eglib/test/path.c [hunk 2]
```diff
@@ -299,6 +299,12 @@ test_cwd (void)
 	char *dir = g_get_current_dir ();
 #ifdef G_OS_WIN32
 	const gchar *newdir = "C:\\Windows";
+#elif defined(_AIX)
+	/*
+	 * AIX/PASE have /bin -> /usr/bin, and chdir follows links.
+	 * Use a directory available on both systems that shouldn't be a link.
+	 */
+	const gchar *newdir = "/usr";
```
Comment (jaykrell) [inline mono/eglib/test/path.c]: How about `/`?

File: mono/eglib/test/test-eglib.exp [hunk 3]
```diff
@@ -0,0 +1,3 @@
+* This file is required so that the symbol is visibile and not GCed on AIX.
+dummy_test_export
+
```
Comment (jaykrell) [inline mono/eglib/test/test-eglib.exp]: Yeah a .def file! :)

File: mono/eglib/test/Makefile.am [hunk 4]
```diff
@@ -37,6 +37,10 @@ test_eglib_SOURCES = $(SOURCES)
 GLIB_TEST_FLAGS_COMMON = -Wall -DEGLIB_TESTS=1 -D_FORTIFY_SOURCE=2 -I$(srcdir)/.. -I.. -DDRIVER_NAME=\"EGlib\"
 CFLAGS += $(GLIB_TEST_FLAGS_COMMON)
 CXXFLAGS += $(GLIB_TEST_FLAGS_COMMON) @CXXFLAGS_COMMON@
+if HOST_AIX
+# This export file is required so test symbols are visible.
```
Comment (jaykrell) [inline mono/eglib/test/Makefile.am]: `so` or `to`?

File: mono/eglib/test/path.c (hunk 2 above)
Comment (NattyNarwhal) [reply mono/eglib/test/path.c]: Fair; it'd probably also make sense to collapse it back down to the general POSIX case too.

File: mono/eglib/test/test-eglib.exp (hunk 3 above)
Comment (NattyNarwhal) [reply mono/eglib/test/test-eglib.exp]: I went with what the AIX linker uses for convention; but it is kinda like that, heh.

File: mono/eglib/test/Makefile.am (hunk 4 above)
Comment (NattyNarwhal) [reply mono/eglib/test/Makefile.am]: It seems grammatically correct to me?

Comment (jaykrell) [reply mono/eglib/test/Makefile.am]: Oops yes.

**Pull request merged**